### PR TITLE
[FIRRTL] Implement MergeConnectionsPass

### PR DESCRIPTION
**Pull request by uenoku**
This commit implements MergeConnectionPass to merge expanded connections into one connection.
LowerTypes fully expands aggregate connections even when semantically
we don't have to expand because it is necessary for ExpandWhen.

More specifically this pass folds the following patterns:
```
%dest(0) <= subfield %a(0)
%dest(1) <= subfield %a(1)
...
%dest(n) <= subfield %a(n)
```
into
```
%dest <= %a
```
This pass also merge connections if their source values are known to be
constant.

With this pass, this fir:
```scala
circuit Bar :
  module Bar :
    input source : {b: UInt<1>, c:UInt<1>}[2][2]
    output sink1 : {b: UInt<1>, c:UInt<1>}[2][2]
    output sink2 : {b: UInt<1>, c:UInt<1>}
    sink1 <= source
    sink2.b <= UInt<1>(1)
    sink2.c <= UInt<1>(1)
```
becomes
```sv
module Bar(     // foo.fir:2:10
  input  struct packed {logic b; logic c; }[1:0][1:0] source,
  output struct packed {logic b; logic c; }[1:0][1:0] sink1,
  output struct packed {logic b; logic c; }           sink2);

  assign sink1 = source;        // foo.fir:2:10
  assign sink2 = /*cast(bit)*/2'h3;     // foo.fir:2:10
endmodule
```

### CODE REVIEW

**Comment by darthscsi** [comment]
> even when semantically
> we don't have to expand because it is necessary for ExpandWhen

Is there a formulation of ExpandWhens where we don't need to expand connects?  It's wasteful to expand and then unexpand.  unexpanding because chisel gives expanded connects is reason enough for the pass, but we should be doing extra work.

File: lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+// be constant.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+#include <queue>
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp:29]: You are just using this for a worklist.  SmallVector (with a reasonable constant) is probably better.

File: lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+
+#define DEBUG_TYPE "firrtl-merge-connections"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+// Return ture if it might be fine to merge the connect.
+static bool okToPeel(ConnectOp connect) {
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp:38]: static functions don't need to be in an anonymous namespace.

File: lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+  // Ignore connections between different types.
+  if (connect.src().getType() != connect.dest().getType())
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp:40]: This isn't strictly necessary, right?  connects don't need pairwise identical types.

File: lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+    return false;
+  // Ignore if either source or destination is an argument.
+  if (!connect.dest().getDefiningOp() || !connect.src().getDefiningOp())
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp:43]: A bundle port should be usable, right?

File: lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+    return false;
+  return true;
+}
+
+// Return the constant value if value is essentially constant.
+static Value getConstant(Value value) {
+  if (auto constant = value.getDefiningOp<ConstantOp>())
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp:50]: op = getDefiningOp();
if (op && isa<ConstantOp, InvalidValueOp, ...>()) ...

File: lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+  // TODO: Add unrealized_conversion, asUInt, asSInt
+  return {};
+}
+
+//===----------------------------------------------------------------------===//
+// Pass Infrastructure
+//===----------------------------------------------------------------------===//
+
+// A helper struct to merge connections.
+struct MergeConnection {
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp:66]: This should be in an anonymous namespace.

File: lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+  // A work list for constant connection merging. We use a queue to traverse
+  // leafs first.
+  std::queue<Value> constantConnectWorklist;
+
+  // Container for unused connections to erase in the post-process.
+  SmallVector<ConnectOp> shouldBeRemoved;
+};
+
+// Peel if the connect might be a part of full connections.
+bool MergeConnection::peelFullyConnect(ConnectOp connect) {
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp:118]: Would it be advantageous to restructure this as in the first pass collecting all connects by source (with map source -> set of connects), then testing each element in the map to see if it has a complete set of connects?

**Review by darthscsi — APPROVED** [review]
This can be merged with minor changes, but I think there's some restructuring which would make things simpler and faster.

**Comment by uenoku** [comment]
> Is there a formulation of ExpandWhens where we don't need to expand connects? It's wasteful to expand and then unexpand. unexpanding because chisel gives expanded connects is reason enough for the pass, but we should be doing extra work.

ExpandWhens works well without extra modification under the condition that connections are expanded *properly*. 
That is, for each connection destination, if any child of the destination is not used as destination, ExpandWhen works  correctly.

File: lib/Dialect/FIRRTL/FIRRTLDialect.cpp [hunk 8]
```diff
@@ -122,6 +122,36 @@ std::string circt::firrtl::getFieldName(const FieldRef &fieldRef,
   return name.str().str();
 }
 
+Type circt::firrtl::getTypeFromFieldRef(FieldRef fieldRef) {
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/FIRRTLDialect.cpp:125]: I'll use https://github.com/llvm/circt/pull/2505 when merged.

File: lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp (hunk 7 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp]: Thanks, I restructured in a similar way.

**Comment by uenoku** [comment]
Refactored to use concat + bitcast. Restructured to visit all connections first. 

Now, we can the get following ouput.   
```scala
circuit test:
  module test:
    output sink: UInt<1>[2],
    input source0: UInt<1>,
    input source1: UInt<1>,

    sink[0] <= source0
    sink[1] <= source1
```
```sv
module test(    // bar.fir:2:10
  input        source0, source1,
  output [1:0] sink);

  assign sink = {source0, source1};     // bar.fir:2:10
endmodule
```

File: include/circt/Dialect/FIRRTL/FIRRTLDialect.h [hunk 9]
```diff
@@ -33,6 +33,9 @@ namespace firrtl {
 /// location.
 FieldRef getFieldRefFromValue(Value value);
 
+/// Get the type specified by fieldRef.
+Type getTypeFromFieldRef(FieldRef fieldRef);
```
Comment (darthscsi) [inline include/circt/Dialect/FIRRTL/FIRRTLDialect.h:37]: Something like this got merge separately, yes?

Comment (darthscsi) [reply include/circt/Dialect/FIRRTL/FIRRTLDialect.h:37]: I see your comment below now.

File: lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+    }
+
+    // Otherwise, we concat all values and cast them into the aggregate type.
+    Value accumulate;
+    changed = true;
+    for (auto value : operands) {
+      value = builder.createOrFold<BitCastOp>(
+          value.getLoc(),
+          UIntType::get(value.getContext(),
+                        *firrtl::getBitWidth(
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp:156]: Did I miss the check that all the components have fixed width and are integer-like?  (at least are legal to concat?)
What about flip types in the aggregate?

File: lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+      continue;
+
+    unsigned size = shouldBeRemoved.size();
+    auto value = getMergedConnectedValue({root, 0});
+    if (!value) {
+      // If we failed to get merged value, revert dead connections.
+      shouldBeRemoved.resize(size);
+      continue;
+    }
+    auto builder = OpBuilder::atBlockEnd(moduleOp.getBody());
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp:211]: This pass assumes expandwhens is done?

**Review by darthscsi — COMMENTED** [review]
It seems like this pass could be done as a single iteration backwards over a module.
You collect connects along the way, when you see a complete connect for a value, then you replace the final connect with the completed connect, then add it to the collection, delete the unnecessary ones, and keep going.  As a bonus, this should make the pass progressive, in that leaf aggregates could be merged, even if the parent aggregate can't.

File: lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp (hunk 11 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp:211]: Yes, definitely. Without expandWhens, this pass is illegal.

File: lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp (hunk 10 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp:156]: I had implicit assumptions that inputs came through inferWidth and LowerTypes with current aggregate preservation.
They guarantee that all preserved aggregate types don't contain analog-type or flip type. I agree that I should check these assumption explicitly so I'll add.

**Comment by uenoku** [comment]
Refactored to do forward iterations. Because of the duality between MergeConnections and LowerTypes, it is very natural to run this pass in the forward iterations.

File: lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp (hunk 10 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/MergeConnections.cpp:184]: Added checks at begging of `peelConnection`

**Comment by uenoku** [comment]
To prevent the pass from creating large aggregate values, I added an option to switch behavior. If `aggressvie-merging` is enabled, the pass merges all the connections. If not, the pass merges only when source values are simplified e.g. constant or simple aggregate value.

**Pull request merged**